service/watcher/configmap/watch.go

```diff
@@ -0,0 +1,155 @@
+package configmap
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/giantswarm/microerror"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"github.com/giantswarm/app-operator/v2/pkg/annotation"
+	pkglabel "github.com/giantswarm/app-operator/v2/pkg/label"
+)
+
+func (c *AppValueWatcher) watch(ctx context.Context) {
+	for {
+		lo := metav1.ListOptions{
+			LabelSelector: pkglabel.Watching,
+		}
+
+		// Find the highest resourceVersion for each configmap.
+		cms, err := c.k8sClient.K8sClient().CoreV1().ConfigMaps("").List(ctx, lo)
+		if err != nil {
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("failed to get configmaps with label %#q", pkglabel.Watching), "stack", fmt.Sprintf("%#v", err))
+			continue
+		}
+
+		var highestResourceVersion uint64
+
+		for _, cm := range cms.Items {
+			currentResourceVersion := getResourceVersion(cm.GetResourceVersion())
+			if highestResourceVersion < currentResourceVersion {
+				highestResourceVersion = currentResourceVersion
+			}
+		}
+
+		c.logger.LogCtx(ctx, "debug", fmt.Sprintf("starting ResourceVersion is %d", highestResourceVersion))
+
+		lo.ResourceVersion = strconv.FormatUint(highestResourceVersion, 10)
+
+		res, err := c.k8sClient.K8sClient().CoreV1().ConfigMaps("").Watch(ctx, lo)
+		if err != nil {
+			panic(err)
+		}
+
+		for r := range res.ResultChan() {
+			if r.Type == watch.Bookmark || r.Type == watch.Error {
+				// no-op for unsupported events
+				continue
+			}
+
+			cm, err := toConfigMap(r.Object)
+			if err != nil {
+				panic(err)
+			}
+
+			configMap := configMapIndex{
+				Name:      cm.GetName(),
+				Namespace: cm.GetNamespace(),
+			}
+
+			var storedIndex map[appIndex]bool
+			{
+				v, ok := c.configMapToApps.Load(configMap)
+				if !ok {
+					c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("cache missed configMap %#q in namespace %#q", configMap.Name, configMap.Namespace))
+					continue
+				}
+
+				storedIndex, ok = v.(map[appIndex]bool)
+				if !ok {
+					panic(fmt.Sprintf("expected '%T', got '%T'", map[appIndex]bool{}, v))
```

Comment:
This panic is also a concern for me. 

What happens if we hit it? Will the operator still reconcile other CRs?

Reply:
The pod would be terminated after hitting this. So far the operator only put the same type in `Store` so I don't think we would ever hit this line. Do you think it's better to have error handling?

Reply:
Agree its very unlikely we'll hit this but I'd still like to avoid the panic.

What about logging it and continuing? If the watcher isn't working but the operator stays running I think that's better.

Reply:
Sure, Thanks for the suggestions, I added error handlings and fix the main function as a goroutine, too. PTAL :)

Reply:
Great those are nice improvements. Thanks for taking care.